### PR TITLE
migrate token pooling to postgres

### PR DESCRIPTION
**Pull request by chris48s**
This PR implements the suggestion from issue #8921 allowing us to migrate the token pool from redis to fly.io postgres.

### CODE REVIEW

File: core/token-pooling/sql-token-persistence.integration.js
Comment (chris48s) [inline core/token-pooling/sql-token-persistence.integration.js:63]: This isn't the most efficient way to insert this data, but given we are only dealing with 3 records.. meh :shrug:

File: package.json
Comment (chris48s) [inline package.json:57]: Even though we don't expect to change our DB schema super often, I think its worth having a migration package to make it easier to evolve this over time. Ideally we should also look at running `npm run migrate up` on deploy. For the first deploy I think it can live with doing it manually.

File: migrations/1676731511125_initialize.cjs
Comment (chris48s) [inline migrations/1676731511125_initialize.cjs:1]: This needs to be a CommonJS file even though our project is ESM, which is why it uses the `.cjs` extension

File: core/token-pooling/sql-token-persistence.js
Comment (chris48s) [inline core/token-pooling/sql-token-persistence.js:28]: `ON CONFLICT (token) DO NOTHING` effectively makes this an "upsert" which closely mirrors the behaviour of a redis `sadd()`

File: package-lock.json
Comment (chris48s) [inline package-lock.json:53151]: enforce-dependency-review is complaining about this. I figure we're going to merge a bunch of dependabot stuff ahead of this and I'm going to need to resolve a conflict on `package-lock.json` anyway. Lets ignore it for now and circle back when we're ready to merge it. I've noted it.

File: core/token-pooling/sql-token-persistence.integration.js
Comment (chris48s) [inline core/token-pooling/sql-token-persistence.integration.js:44]: This copies our table to a temp schema for the purposes of running the tests then clears it up again after. This means we can run the tests against a local DB with real data in it and not break our local data. We don't need a separate test DB. It also means the tests will keep step with schema changes from the migrations. We do need to `migrate up` before running the tests though.

File: package.json
Comment (chris48s) [inline package.json:123]: This is a bit hacky. I want the DB config to be specify-able in yaml or env vars like all our other setting but `node-pg-migrate` has its own ideas about how to handle config https://salsita.github.io/node-pg-migrate/#/cli . This little script reads the postgres connection string (however it has been set - yaml or env vat) and writes it to a config file for `node-pg-migrate`. I did consider writing a script that outputs the connection string to stdout and then setting an env var with `DATABASE_URL=$(node /scripts/get-dataabse-url.js)` but I picked this way because it should be cross-platform for Windows users.

File: server.js
Comment (chris48s) [inline server.js:50]: I decided that it is useful to keep both backends around for some length of time because it allows us to switch between them just by changing an env var - no code changes needed. Doing this means we can:

- deploy this code with `REDIS_URL` set (so we're still using the redis backend)
- migrate our data to postgres with `scripts/migrate-token-pool.js`
- set `POSTGRES_URL` to switch to the postgres backend
- if it all goes wrong, we can easily just unset `POSTGRES_URL` to go back to redis
- unset `REDIS_URL` once we're happy everything is fine
- delete the redis code at a later date

This both makes the migration super low-risk/easy to back out of and means we can provide a migration path for self-hosting users with basically no effort.

File: scripts/migrate-token-pool.js
Comment (chris48s) [inline scripts/migrate-token-pool.js:8]: For this script it is useful to pass the connection strings in on the console instead of using the application config because it allows us to copy the data over before we set `POSTGRES_URL` to switch the pool over.

Also I'm not too worried about achieving perfection in this specific script. It is a file we are going to use once and then delete it.

File: core/token-pooling/sql-token-persistence.integration.js
Comment (chris48s) [inline core/token-pooling/sql-token-persistence.integration.js:99]: The test cases in this file are the exact same ones we have for redis.

File: server.js
Comment (calebcartwright) [reply server.js:50]: Might be worth extending the warning message to include a link to https://github.com/badges/shields/issues/8921

File: core/token-pooling/sql-token-persistence.integration.js
Comment (calebcartwright) [reply core/token-pooling/sql-token-persistence.integration.js:44]: > This means we can run the tests against a local DB with real data in it and not break our local data. We don't need a separate test DB. It also means the tests will keep step with schema changes from the migrations. We do need to migrate up before running the tests though

No objections to this in the local/inner dev loop (nor CI) context, and agree the benefits outweigh the tradeoffs. Would be nervous abut this in the context of the real DB though (whether connected to deliberately or accidentally)

File: migrations/1676731511125_initialize.cjs
Comment (calebcartwright) [inline migrations/1676731511125_initialize.cjs:3]: What's this?

File: package.json
Comment (calebcartwright) [reply package.json:57]: Does this need to be a prod dep or can it still be utilized as a dev dep?

**Comment by calebcartwright** [comment]
LGTM, happy to :+1: once merge conflicts are sorted

File: package.json
Comment (chris48s) [reply package.json:57]: In order to create the DB (in the first instance) or modify it (later) we need to be able to run `npm run migrate up` in the production environment so it has to be a prod dependency.

File: core/token-pooling/sql-token-persistence.integration.js
Comment (chris48s) [reply core/token-pooling/sql-token-persistence.integration.js:44]: In principle it should be safe but I agree - I wouldn't want to run the tests while connected to the prod DB. Tbh that would be fairly unusual thing to do for any application. I think the fact that the dev dependencies (e.g: mocha, etc) aren't installed in the prod container is enough of a guard rail to stop this from happening.

Personally I'd tend to avoid connecting my dev environment up to a prod database for any application.

File: server.js
Comment (chris48s) [reply server.js:50]: Yes. I'll call this out at the top of the changelog for the next server release and write it with a bit more of a focus on self-hosting users than the discussion in https://github.com/badges/shields/issues/8921 (which is quite shields.io-specific). I'm going to set myself a reminder to ping me so I remember to do this when I do the changelog.

@SnoozeThis wait until Feb 28 2023

File: migrations/1676731511125_initialize.cjs
Comment (chris48s) [reply migrations/1676731511125_initialize.cjs:3]: https://github.com/salsita/node-pg-migrate/blob/9331f6fda98795e3f3733461f9b71345168b99d8/docs/migrations.md?plain=1#L3-L19

**Comment by chris48s** [comment]
@SnoozeThis wait until Feb 28 2023

**Comment by SnoozeThis** [comment]
(https://snoozeth.is/0ky67I-_i6g) I will wait until Tue, 28 Feb 2023 00:00:00 UTC and then add a comment.

**Comment by chris48s** [comment]
Should be good to go, I think.

I can look at setting up a postgres instance on fly and getting the actual data migration done in the week.

**Review by calebcartwright — APPROVED** [review]

**Comment by SnoozeThis** [comment]
Resolved.

**Pull request merged**
